From e1d404ddd5fb77ac408437a44e11fa3e894fe5c3 Mon Sep 17 00:00:00 2001
From: Michael Webster <miketwebster@gmail.com>
Date: Tue, 8 Nov 2016 19:58:09 -0500
Subject: [PATCH] unlock/accountsServiceClient: don't cache the username and
 face path, instead fetch them from AccountsService when used.  Also, allow
 for no face image by not using a fallback generic image, and hiding the image
 widget if there is no face for it.

---
 src/dbusdepot/accountsServiceClient.py | 21 ++++++++++++---------
 src/unlock.py                          | 11 +++++++----
 2 files changed, 19 insertions(+), 13 deletions(-)

diff --git a/src/dbusdepot/accountsServiceClient.py b/src/dbusdepot/accountsServiceClient.py
index 8adda63..0e45056 100644
--- a/src/dbusdepot/accountsServiceClient.py
+++ b/src/dbusdepot/accountsServiceClient.py
@@ -20,8 +20,6 @@ def __init__(self):
         super(AccountsServiceClient, self).__init__()
 
         self.is_loaded = False
-        self.real_name = None
-        self.face_path = None
 
         self.service = AccountsService.UserManager.get_default().get_user(utils.get_user_name())
         trackers.con_tracker_get().connect(self.service,
@@ -29,14 +27,19 @@ def __init__(self):
                                            self.on_accounts_service_loaded)
 
     def on_accounts_service_loaded(self, service, param):
-        self.real_name = service.get_real_name()
+        self.is_loaded = True
+        self.emit("account-loaded")
+
+    def get_real_name(self):
+        return self.service.get_real_name()
 
-        for path in [os.path.join(service.get_home_dir(), ".face"),
-                     service.get_icon_file(),
-                     "/usr/share/cinnamon/faces/user-generic.png"]:
+    def get_face_path(self):
+        face_path = None
+
+        for path in [os.path.join(self.service.get_home_dir(), ".face"),
+                     self.service.get_icon_file()]:
             if os.path.exists(path):
-                self.face_path = path
+                face_path = path
                 break
 
-        self.is_loaded = True
-        self.emit("account-loaded")
+        return face_path
\ No newline at end of file
diff --git a/src/unlock.py b/src/unlock.py
index 6e22d9a..4f24e9b 100644
--- a/src/unlock.py
+++ b/src/unlock.py
@@ -51,6 +51,8 @@ def __init__(self):
         self.face_image = FramedImage()
         self.face_image.set_halign(Gtk.Align.CENTER)
         self.face_image.get_style_context().add_class("faceimage")
+        self.face_image.set_no_show_all(True)
+
         self.box.pack_start(self.face_image, False, False, 10)
 
         self.realname_label = Gtk.Label(None)
@@ -177,12 +179,13 @@ def on_account_client_loaded(self, client):
         """
         Handler for the AccountsService - requests the user real name and .face image.
         """
-        if client.real_name != None:
-            self.real_name = client.real_name
+        if client.get_real_name() != None:
+            self.real_name = client.get_real_name()
             self.update_realname_label()
 
-        if client.face_path != None:
-            self.face_image.set_from_path(client.face_path)
+        if client.get_face_path() != None:
+            self.face_image.set_from_path(client.get_face_path())
+            self.face_image.show()
 
     def on_password_entry_text_changed(self, editable):
         """
